chain/src/txhashset.rs

```diff
@@ -1072,14 +1071,76 @@ pub fn zip_read(root_dir: String) -> Result<File, Error> {
 
 /// Extract the txhashset data from a zip file and writes the content into the
 /// txhashset storage dir
-pub fn zip_write(root_dir: String, txhashset_data: File) -> Result<(), Error> {
+pub fn zip_write(root_dir: String, txhashset_data: File, header: &BlockHeader) -> Result<(), Error> {
 	let txhashset_path = Path::new(&root_dir).join(TXHASHSET_SUBDIR);
+	check_files(&txhashset_path, header)?;
 
 	fs::create_dir_all(txhashset_path.clone())?;
 	zip::decompress(txhashset_data, &txhashset_path)
 		.map_err(|ze| ErrorKind::Other(ze.to_string()).into())
 }
 
+/// Check that the txhashset directory does not contains unexpected files
+fn check_files(txhashset_path: &PathBuf, header: &BlockHeader) -> Result<(), Error> {
+	let subdirectories_expected: HashSet<_> = [OUTPUT_SUBDIR, KERNEL_SUBDIR, RANGE_PROOF_SUBDIR]
+		.iter()
+		.cloned()
+		.map(|s| String::from(s))
+		.collect();
+	let subdirectories_found: HashSet<_> = fs::read_dir(txhashset_path)
+		.unwrap()
+		.filter_map(|entry| {
+			entry.ok().and_then(|e| {
+				e.path()
+					.file_name()
+					.and_then(|n| n.to_str().map(|s| String::from(s)))
+			})
+		})
+		.collect();
+	let dir_difference: Vec<String> = subdirectories_found
+		.difference(&subdirectories_expected)
+		.cloned()
+		.collect();
+	if !dir_difference.is_empty() {
+		return Err(ErrorKind::Other(
+			"Unexpected file(s) found in txhashset folder".to_string(),
+		).into());
+	}
+
+	// Checking now every subdirectory
+	let pmmr_files_expected: HashSet<_> = PMMR_FILES
+		.iter()
+		.cloned()
+		.map(|s| if s.contains(PMMR_FILES[2]) {
+		format!("{}.{}", s, header.hash())}
+		else {String::from(s) })
+		.collect();
+	let subdirectories = fs::read_dir(txhashset_path).unwrap();
+	for subdirectory in subdirectories {
+		let subdirectory_path = subdirectory.unwrap().path();
+		let pmmr_files = fs::read_dir(subdirectory_path).unwrap();
```

Comment:
this code looks like a copy code above starting from line 1090, can we make euse it somehow?